x/mongo/driver/operation.go

```diff
@@ -1713,76 +1740,168 @@ func (op *Operation) redactCommand(cmd string, doc bsoncore.Document) bool {
+			copy(cmdCopy, info.cmd)
+			if info.documentSequenceIncluded {
+				cmdCopy = cmdCopy[:len(info.cmd)-1] // remove 0 byte at end
+				cmdCopy = op.addBatchArray(cmdCopy)
+				cmdCopy, _ = bsoncore.AppendDocumentEnd(cmdCopy, 0) // add back 0 byte and update length
+			}
 		}
+		return cmdCopy
+	}
+
+	// If logging is enabled for the command component at the debug level, log the command response.
+	if op.canLogCommandMessage() {
+		host, port, _ := net.SplitHostPort(info.serverAddress.String())
+
+		op.Logger.Print(logger.LevelDebug, &logger.CommandStartedMessage{
+			Command:      getCmdCopy(),
+			DatabaseName: op.Database,
+
+			CommandMessage: logger.CommandMessage{
+				MessageLiteral:     logger.CommandMessageStartedDefault,
+				Name:               info.cmdName,
+				RequestID:          int64(info.requestID),
+				ServerConnectionID: info.serverConnID,
+				ServerHost:         host,
+				ServerPort:         port,
+				ServiceID:          info.serviceID,
+			},
+		})
 	}
 
-	started := &event.CommandStartedEvent{
-		Command:            cmdCopy,
-		DatabaseName:       op.Database,
-		CommandName:        info.cmdName,
-		RequestID:          int64(info.requestID),
-		ConnectionID:       info.connID,
-		ServerConnectionID: info.serverConnID,
-		ServiceID:          info.serviceID,
+	if op.canPublishStartedEven() {
+		started := &event.CommandStartedEvent{
+			Command:            getCmdCopy(),
+			DatabaseName:       op.Database,
+			CommandName:        info.cmdName,
+			RequestID:          int64(info.requestID),
+			ConnectionID:       info.connID,
+			ServerConnectionID: info.serverConnID,
+			ServiceID:          info.serviceID,
+		}
+		op.CommandMonitor.Started(ctx, started)
 	}
-	op.CommandMonitor.Started(ctx, started)
+}
+
+// canPublishSucceededEvent returns true if a CommandSucceededEvent can be
+// published for the given command. This is true if the command is not an
+// unacknowledged write and the command monitor is monitoring succeeded events.
+func (op Operation) canPublishFinishedEvent(info finishedInformation) bool {
+	success := info.success()
+
+	return op.CommandMonitor != nil &&
+		(!success || op.CommandMonitor.Succeeded != nil) &&
+		(success || op.CommandMonitor.Failed != nil)
 }
 
 // publishFinishedEvent publishes either a CommandSucceededEvent or a CommandFailedEvent to the operation's command
 // monitor if possible. If success/failure events aren't being monitored, no events are published.
 func (op Operation) publishFinishedEvent(ctx context.Context, info finishedInformation) {
-	success := info.cmdErr == nil
-	if _, ok := info.cmdErr.(WriteCommandError); ok {
-		success = true
+	// rawResponse is the raw response from the server.
+	var rawResponse bson.Raw
+
+	// getRawResponse is a closure that returns the raw response from the server. It is used to lazy load the
+	// rawResponse variable.
+	var getRawResponse = func() bson.Raw {
+		if rawResponse != nil {
+			return rawResponse
```

Comment:
Again, I worry this line will use the same response copy for logging and monitoring which may cause odd behavior.